packages/drivers/routerlicious-driver/src/sessionInfoManager.ts

```diff
@@ -0,0 +1,145 @@
+ * Currently, R11s defines session length at 10 minutes. To avoid any weird unknown edge-cases though,
+ * we set the limit to 5 minutes here.
+ * In the future, we likely want to retrieve this information from service's "inactive session" definition.
+ */
+export const RediscoverAfterTimeSinceDiscoveryMs = 5 * 60000; // 5 minutes
+
+interface IGetSessionInfoParams {
+	resolvedUrl: IResolvedUrl;
+	documentId: string;
+	tenantId: string;
+	ordererRestWrapper: RouterliciousOrdererRestWrapper;
+	logger: ITelemetryLoggerExt;
+}
+
+export interface IGetSessionInfoResponse {
+	refreshed: boolean;
+	resolvedUrl: IResolvedUrl;
+}
+
+export class SessionInfoManager {
+	/**
+	 * Stored session info
+	 * Key: URL for given session (see "getDiscoverSessionUrl")
+	 * Value: session info stored as an IResolvedUrl
+	 */
+	private readonly sessionFluidUrls: Map<string, IResolvedUrl> = new Map();
+
+	/**
+	 * Stored dates of when a session was last discovered/refreshed
+	 * Key: URL for given session (see "getDiscoverSessionUrl")
+	 * Value: date last discovered
+	 */
+	private readonly sessionLastDiscovered: Map<string, number> = new Map();
+
+	constructor(private readonly enableDiscovery: boolean) {}
+
+	/**
+	 * Start tracking info for a given session
+	 */
+	public async initializeSessionInfo(
+		params: IGetSessionInfoParams & { session: ISession | undefined },
+	): Promise<IResolvedUrl> {
+		const { resolvedUrl, session } = params;
+
+		const url = getDiscoverSessionUrl(params);
+		assert(
+			this.sessionFluidUrls.has(url) === this.sessionLastDiscovered.has(url),
+			"Session map state mismatch",
+		);
+
+		if (session !== undefined) {
+			this.sessionFluidUrls.set(url, getDiscoveredFluidResolvedUrl(resolvedUrl, session));
+			this.sessionLastDiscovered.set(url, Date.now());
+		} else if (!this.sessionFluidUrls.has(url)) {
+			this.sessionFluidUrls.set(url, resolvedUrl);
+			// Force a refresh
+			this.sessionLastDiscovered.set(url, 0);
+		}
+
+		return (await this.getSessionInfo(params)).resolvedUrl;
+	}
+
+	/**
+	 * Retrieve, and potentially refresh, info of a given session
+	 */
+	public async getSessionInfo(
+		params: IGetSessionInfoParams,
+	): Promise<IGetSessionInfoResponse> {
+		const url = getDiscoverSessionUrl(params);
+		assert(
+			this.sessionFluidUrls.has(url) && this.sessionLastDiscovered.has(url),
+			"Unexpected discover session URL",
+		);
+
+		let refreshed = false;
+		const shouldRediscover =
+			Date.now() - this.sessionLastDiscovered.get(url)! > RediscoverAfterTimeSinceDiscoveryMs;
+		if (this.enableDiscovery && shouldRediscover) {
+			await this.updateSessionInfo(params).catch(() => {
+				// Undo discovery time set on failure, so that next check refreshes.
+				this.sessionLastDiscovered.set(url, 0);
+			});
+			refreshed = true;
+		}
+		return {
+			refreshed,
+			resolvedUrl: this.sessionFluidUrls.get(url)!,
+		};
+	}
+
+	private async updateSessionInfo(params: IGetSessionInfoParams): Promise<void> {
```

Comment:
>updateSessionInfo [](http://example.com/codeflow?start=15&length=17)

nit: `fetchSessionInfo` or `fetchFreshSessionInfo` will be more explicit in depicting that we are making the Discover Session call as opposed to using a cached value from the map. Helps readability